### PR TITLE
FAI-478 - more accurate data generation for feature intervals

### PR DESCRIPTION
**Pull request by tteofili**
We improve the way feature intervals are currently generated for tabular data when a (small) data distribution is provided to `LimeExplainer` (via `LimeConfig#withDataDistribution`).
We filter out points generated using _bootstrap_ that are associated to "low score zones", as far as the model to be explained is concerned. In fact such points are unlikely to be handled correctly by the model and hence produce out of distribution samples (that are then used to generate the LIME explanation).

See https://issues.redhat.com/browse/FAI-478

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

**Comment by tteofili** [comment]
Jenkins retest this please

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java [hunk 1]
```diff
@@ -576,9 +586,17 @@ public static DataDistribution readCSV(Path file, List<Type> schema) throws IOEx
...
+                    // only use high score points, if possible
+                    HighScoreNumericFeatureZones highScoreNumericFeatureZones = numericFeatureZonesMap.get(feature.getName());
+                    if (highScoreNumericFeatureZones != null) {
+                        data = Arrays.stream(data).filter(highScoreNumericFeatureZones::accept).toArray();
+                    }
+
+                    double mean = DataUtils.getMean(data);
                     double stdDev = Math
-                            .pow(DataUtils.getStdDev(sampledValues.stream().mapToDouble(Value::asNumber).toArray(), mean), 2);
+                            .pow(DataUtils.getStdDev(data, mean), 2);
```
Comment (r00ta) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java]: Is it possible that `Arrays.stream(data).filter(highScoreNumericFeatureZones::accept)` produces an array with 0 elements? 
If so, I think `DataUtils.getMean(data)` will throw an exception (`0/0`) since the implementation is 
```
    public static double getMean(double[] data) {
        double m = 0;
        for (double datum : data) {
            m += datum;
        }
        m = m / data.length;
        return m;
    }
```

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        PredictionProvider predictionProvider = TestUtils.getSumThresholdModel(0.1, 0.1);
+        List<FeatureDistribution> featureDistributions = new ArrayList<>();
+        DataDistribution dataDistribution = new IndependentFeaturesDataDistribution(featureDistributions);
+        Map<String, HighScoreNumericFeatureZones> highScoreFeatureZones =
+                HighScoreNumericFeatureZonesProvider.getHighScoreFeatureZones(dataDistribution, predictionProvider, features);
+        assertThat(highScoreFeatureZones).isNotNull();
+        assertThat(highScoreFeatureZones.size()).isEqualTo(0);
+    }
+
+}
```
Comment (r00ta) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java]: Add test for non-empty data?

**Review by r00ta — APPROVED** [review]
I can't say much about the algorithm itself since I'm not an expert on this. 
I just have a couple of small comments, LGTM :+1:

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java (hunk 2 above)
Comment (tteofili) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java]: good point, thx

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java (hunk 1 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java]: correct, thanks ! I'll make sure the data array doesn't get emptied.

**Review by jiripetrlik — APPROVED** [review]
LGTM

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/DataDistribution.java [hunk 3]
```diff
@@ -49,4 +49,11 @@
      * @return a list of feature distributions
      */
     List<FeatureDistribution> asFeatureDistributions();
+
+    /**
+     * Get the size of this data distribution.
+     * 
+     * @return the no. of samples in this data distribution.
+     */
+    int size();
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/DataDistribution.java]: If I read this javadoc I understand that this method is equivalent to `getAllSamples().size()` but then the impl of this method is not really "imposing" this to be true and in [`IndependentFeaturesDataDistribution`](https://github.com/kiegroup/kogito-apps/pull/1015/files#diff-21efd81900a8b11448d2573821050474d6e30bc56a94959ce7779314c65b01c5R101-R108) the logic is quite different so it could be easily come different in the future.
Am I correct?

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZones.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    public HighScoreNumericFeatureZones(double[] highFeatureScorePoint, double tolerance) {
+        this.highFeatureScorePoint = highFeatureScorePoint;
+        this.tolerance = tolerance;
+    }
+
+    public boolean accept(double point) {
+        return DoubleStream.of(highFeatureScorePoint).anyMatch(d -> point > (d - tolerance) && point < (d + tolerance));
+    }
+}
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZones.java]: Do you expect this class to have additional behaviours in the future? Because at the moment it looks like a `Predicate`

```suggestion
public class HighScoreNumericFeatureZones implements Predicate<Double> {

    private final double[] highFeatureScorePoint;
    private final double tolerance;

    public HighScoreNumericFeatureZones(double[] highFeatureScorePoint, double tolerance) {
        this.highFeatureScorePoint = highFeatureScorePoint;
        this.tolerance = tolerance;
    }

    @Override
    public boolean test(Double point) {
        return DoubleStream.of(highFeatureScorePoint).anyMatch(d -> point > (d - tolerance) && point < (d + tolerance));
    }
}
```

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/LimeExplainer.java [hunk 5]
```diff
@@ -317,19 +318,28 @@ private double toDouble(Output output, Value fv) {
...
-        // generate feature distributions, if possible
-        Map<String, FeatureDistribution> featureDistributionsMap = DataUtils.boostrapFeatureDistributions(
-                limeConfig.getDataDistribution(), perturbationContext, 2 * size,
-                1, size);
+        DataDistribution dataDistribution = limeConfig.getDataDistribution();
+
+        Map<String, FeatureDistribution> featureDistributionsMap;
+        if (dataDistribution.size() > 0) {
+            Map<String, HighScoreNumericFeatureZones> numericFeatureZonesMap = HighScoreNumericFeatureZonesProvider
+                    .getHighScoreFeatureZones(dataDistribution, predictionProvider, features);
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/LimeExplainer.java]: This change can have a huge impact on the cost of `LimeExplainer` (internally it will calculate the prediction using all available samples in the distribution), I think we can make it opt-in or at least adding a log message.
Wdyt?

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProvider.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            Thread.currentThread().interrupt();
+        } catch (TimeoutException e) {
+            LOGGER.error("Timed out while waiting for sorting predictions by score", e);
+        }
+        if (!scoreSortedPredictions.isEmpty()) {
+            // calculate min, max and mean scores
+            double max = scoreSortedPredictions.get(0).getOutput().getOutputs().stream().mapToDouble(Output::getScore).sum();
+            double min = scoreSortedPredictions.get(scoreSortedPredictions.size() - 1).getOutput().getOutputs().stream()
+                    .mapToDouble(Output::getScore).sum();
+            double threshold = (max + min) / 2;
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProvider.java]: Why `threshold` instead of `meanScore` (like in the code comment) or similar?

**Review by danielezonca — COMMENTED** [review]
Overall looks good to me, just some clarification

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZones.java (hunk 4 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZones.java]: thanks, it makes sense to convert it to a `Predicate`

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/LimeExplainer.java (hunk 5 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/LimeExplainer.java]: good point, I'll add a config option.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/DataDistribution.java (hunk 3 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/DataDistribution.java]: good point, however I've realized what is needed is just `distribution.isEmpty()`, which is simpler to implement and is not tied to any `sample` calls.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProvider.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            LOGGER.error("Timed out while waiting for sorting predictions by score", e);
+        }
+        if (!scoreSortedPredictions.isEmpty()) {
+            // calculate min, max and mean scores
+            double max = scoreSortedPredictions.get(0).getOutput().getOutputs().stream().mapToDouble(Output::getScore).sum();
+            double min = scoreSortedPredictions.get(scoreSortedPredictions.size() - 1).getOutput().getOutputs().stream()
+                    .mapToDouble(Output::getScore).sum();
+            if (max != min) {
+                double threshold = scoreSortedPredictions.stream().map(
+                        p -> p.getOutput().getOutputs().stream().map(Output::getScore).mapToDouble(d -> d).sum())
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProvider.java]:
```suggestion
                        p -> p.getOutput().getOutputs().stream().mapToDouble(Output::getScore).sum())
```

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/IndependentFeaturesDataDistribution.java [hunk 8]
```diff
@@ -97,4 +97,20 @@ public PredictionInput sample() {
...
+            for (FeatureDistribution featureDistribution : this.featureDistributions) {
+                if (!featureDistribution.isEmpty()) {
+                    empty = false;
+                    break;
+                } else {
+                    empty = true;
+                }
+            }
+        }
+        return empty;
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/IndependentFeaturesDataDistribution.java]: What about this alternative? I find it easier to understand 🤔 
```suggestion
        if (this.featureDistributions.isEmpty()) {
            return true;
        }
        for (FeatureDistribution featureDistribution : this.featureDistributions) {
            if (!featureDistribution.isEmpty()) {
                return false;
            }
        }
        return true;
```

File: explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/FraudScoringDmnLimeExplainerTest.java [hunk 9]
```diff
@@ -158,6 +158,31 @@ void testExplanationStabilityWithOptimization() throws ExecutionException, Inter
...
+    @Test
+    void testExplanationWithDataDistribution() throws ExecutionException, InterruptedException, TimeoutException {
+        PredictionProvider model = getModel();
+
+        List<PredictionInput> samples = DmnTestUtils.randomFraudScoringInputs();
+        List<PredictionInput> inputs = samples.subList(0, 10);
+
+        Random random = new Random();
+        random.setSeed(0);
+        PerturbationContext perturbationContext = new PerturbationContext(random, 1);
```
Comment (danielezonca) [inline explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/FraudScoringDmnLimeExplainerTest.java:169]: Minor note: if this other PR https://github.com/kiegroup/kogito-apps/pull/1000 is merged before this one we can simplify this code as
```suggestion
        PerturbationContext perturbationContext = new PerturbationContext(0, new Random(), 1);
```

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        DataDistribution dataDistribution = new IndependentFeaturesDataDistribution(featureDistributions);
+        Map<String, HighScoreNumericFeatureZones> highScoreFeatureZones =
+                HighScoreNumericFeatureZonesProvider.getHighScoreFeatureZones(dataDistribution, predictionProvider, features, 10);
+        assertThat(highScoreFeatureZones).isNotNull();
+        assertThat(highScoreFeatureZones.size()).isEqualTo(0);
+    }
+
+    @Test
+    void testNonEmptyData() {
+        PerturbationContext perturbationContext = new PerturbationContext(new Random(), 1);
```
Comment (danielezonca) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java]: You should set the seed (even if as far as I see the test is not really related with actual values)

**Review by danielezonca — APPROVED** [review]
LGTM, some minor comments but I think the PR is ready 👍

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/IndependentFeaturesDataDistribution.java (hunk 8 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/IndependentFeaturesDataDistribution.java]: I have a taste against multiple return statements :-) 
what about:
```
    @Override
    public boolean isEmpty() {
        return this.featureDistributions.isEmpty() || featureDistributions.stream().allMatch(FeatureDistribution::isEmpty);
    }
```

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java (hunk 10 above)
Comment (tteofili) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/HighScoreNumericFeatureZonesProviderTest.java]: ok

**Comment by tteofili** [comment]
Jenkins retest this please

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/IndependentFeaturesDataDistribution.java (hunk 8 above)
Comment (danielezonca) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/IndependentFeaturesDataDistribution.java]: Fine for me :)

**Comment by tteofili** [comment]
Jenkins retest this please
